csrc/multidevice/utils.cpp

```diff
@@ -142,19 +142,49 @@ int64_t numDeviceDims(TensorView* tv) {
 }
 
 bool isResharding(Expr* expr) {
-  std::unordered_set<TensorView*> tvs;
-  for (auto tv : ir_utils::filterByType<TensorView>(expr->inputs())) {
-    tvs.insert(tv);
-  }
-  for (auto tv : ir_utils::filterByType<TensorView>(expr->outputs())) {
-    tvs.insert(tv);
-  }
-  if (tvs.empty()) {
-    return false;
+  // we don't use getTvsWithDifferentSharding because it creates a computeAtMap,
+  // which is too costly
+  for (auto producer : ir_utils::filterByType<TensorView>(expr->inputs())) {
+    for (auto consumer : ir_utils::filterByType<TensorView>(expr->outputs())) {
+      // exit early in the unsharded case for performance
```

Comment:
The loop body has a mix of `continue` and `return true`, which is a bit hard to follow. I'd create a helper function `isResharding(input_tv, output_tv)` and make the loop here simply 

```
for each input_tv {
  for each output_tv {
    if isResharding(input_tv, output_tv)
      return true;
  }
}
return false;
```

Reply:
The problem I see with this suggestion is that the auxialiary function `isResharding(input_tv, output_tv)` needs to know that I/O are consumer/producer, but there is no way to assert that. Should I write such an auxialiary function relying on this implicit assumption?

Reply:
> there is no easy way to assert that

Something like https://github.com/NVIDIA/Fuser/blob/b020415858caa38a77b3c309566903f3fbdbac3d/csrc/root_domain_map.cpp#L61-L66, which is already asserted by PairwiseRootDomainMap?

Reply:
Right, nice! I did as you suggested, I introduced a new function `bool haveDifferentShardings(TensorView* producer, TensorView* consumer)` in `multidevice/utils.(h|cpp)` and now `isResharding(Expr*)` calls this function by looping over the expr's I/O